// genkit/lib/evaluator.js
"use strict";
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var evaluator_exports = {};
__export(evaluator_exports, {
  BaseDataPointSchema: () => import_evaluator.BaseDataPointSchema,
  BaseEvalDataPointSchema: () => import_evaluator.BaseEvalDataPointSchema,
  EvalResponseSchema: () => import_evaluator.EvalResponseSchema,
  EvalResponsesSchema: () => import_evaluator.EvalResponsesSchema,
  EvaluatorInfoSchema: () => import_evaluator.EvaluatorInfoSchema,
  ScoreSchema: () => import_evaluator.ScoreSchema,
  evaluatorRef: () => import_evaluator.evaluatorRef
});
module.exports = __toCommonJS(evaluator_exports);
var import_evaluator = require("@genkit-ai/ai/evaluator");
// Annotate the CommonJS export names for ESM import in node:
0 && (module.exports = {
  BaseDataPointSchema,
  BaseEvalDataPointSchema,
  EvalResponseSchema,
  EvalResponsesSchema,
  EvaluatorInfoSchema,
  ScoreSchema,
  evaluatorRef
});
