// @genkit-ai/ai/lib/formats/index.js
"use strict";
var __defProp = Object.defineProperty;
var __defProps = Object.defineProperties;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropDescs = Object.getOwnPropertyDescriptors;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __getOwnPropSymbols = Object.getOwnPropertySymbols;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __propIsEnum = Object.prototype.propertyIsEnumerable;
var __defNormalProp = (obj, key, value) => key in obj ? __defProp(obj, key, { enumerable: true, configurable: true, writable: true, value }) : obj[key] = value;
var __spreadValues = (a, b) => {
  for (var prop in b || (b = {}))
    if (__hasOwnProp.call(b, prop))
      __defNormalProp(a, prop, b[prop]);
  if (__getOwnPropSymbols)
    for (var prop of __getOwnPropSymbols(b)) {
      if (__propIsEnum.call(b, prop))
        __defNormalProp(a, prop, b[prop]);
    }
  return a;
};
var __spreadProps = (a, b) => __defProps(a, __getOwnPropDescs(b));
var __objRest = (source, exclude) => {
  var target = {};
  for (var prop in source)
    if (__hasOwnProp.call(source, prop) && exclude.indexOf(prop) < 0)
      target[prop] = source[prop];
  if (source != null && __getOwnPropSymbols)
    for (var prop of __getOwnPropSymbols(source)) {
      if (exclude.indexOf(prop) < 0 && __propIsEnum.call(source, prop))
        target[prop] = source[prop];
    }
  return target;
};
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var __async = (__this, __arguments, generator) => {
  return new Promise((resolve, reject) => {
    var fulfilled = (value) => {
      try {
        step(generator.next(value));
      } catch (e) {
        reject(e);
      }
    };
    var rejected = (value) => {
      try {
        step(generator.throw(value));
      } catch (e) {
        reject(e);
      }
    };
    var step = (x) => x.done ? resolve(x.value) : Promise.resolve(x.value).then(fulfilled, rejected);
    step((generator = generator.apply(__this, __arguments)).next());
  });
};
var formats_exports = {};
__export(formats_exports, {
  DEFAULT_FORMATS: () => DEFAULT_FORMATS,
  configureFormats: () => configureFormats,
  defineFormat: () => defineFormat,
  injectInstructions: () => injectInstructions,
  resolveFormat: () => resolveFormat,
  resolveInstructions: () => resolveInstructions
});
module.exports = __toCommonJS(formats_exports);
var import_array = require("./array.js");
var import_enum = require("./enum.js");
var import_json = require("./json.js");
var import_jsonl = require("./jsonl.js");
var import_text = require("./text.js");
function defineFormat(registry, options, handler) {
  const _a = options, { name } = _a, config = __objRest(_a, ["name"]);
  const formatter = { config, handler };
  registry.registerValue("format", name, formatter);
  return formatter;
}
function resolveFormat(registry, outputOpts) {
  return __async(this, null, function* () {
    if (!outputOpts) return void 0;
    if (outputOpts.schema && !outputOpts.format) {
      return registry.lookupValue("format", "json");
    }
    if (outputOpts.format) {
      return registry.lookupValue("format", outputOpts.format);
    }
    return void 0;
  });
}
function resolveInstructions(format, schema, instructionsOption) {
  if (typeof instructionsOption === "string") return instructionsOption;
  if (instructionsOption === false) return void 0;
  if (!format) return void 0;
  return format.handler(schema).instructions;
}
function injectInstructions(messages, instructions) {
  if (!instructions) return messages;
  if (messages.find(
    (m2) => m2.content.find(
      (p) => {
        var _a, _b;
        return ((_a = p.metadata) == null ? void 0 : _a.purpose) === "output" && !((_b = p.metadata) == null ? void 0 : _b.pending);
      }
    )
  )) {
    return messages;
  }
  const newPart = {
    text: instructions,
    metadata: { purpose: "output" }
  };
  let targetIndex = messages.findIndex((m2) => m2.role === "system");
  if (targetIndex < 0)
    targetIndex = messages.map((m2) => m2.role).lastIndexOf("user");
  if (targetIndex < 0) return messages;
  const m = __spreadProps(__spreadValues({}, messages[targetIndex]), {
    content: [...messages[targetIndex].content]
  });
  const partIndex = m.content.findIndex(
    (p) => {
      var _a, _b;
      return ((_a = p.metadata) == null ? void 0 : _a.purpose) === "output" && ((_b = p.metadata) == null ? void 0 : _b.pending);
    }
  );
  if (partIndex > 0) {
    m.content.splice(partIndex, 1, newPart);
  } else {
    m.content.push(newPart);
  }
  const outMessages = [...messages];
  outMessages.splice(targetIndex, 1, m);
  return outMessages;
}
const DEFAULT_FORMATS = [
  import_json.jsonFormatter,
  import_array.arrayFormatter,
  import_text.textFormatter,
  import_enum.enumFormatter,
  import_jsonl.jsonlFormatter
];
function configureFormats(registry) {
  for (const format of DEFAULT_FORMATS) {
    defineFormat(
      registry,
      __spreadValues({ name: format.name }, format.config),
      format.handler
    );
  }
}
// Annotate the CommonJS export names for ESM import in node:
0 && (module.exports = {
  DEFAULT_FORMATS,
  configureFormats,
  defineFormat,
  injectInstructions,
  resolveFormat,
  resolveInstructions
});
