// @genkit-ai/core/lib/index.js
"use strict";
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __reExport = (target, mod, secondTarget) => (__copyProps(target, mod, "default"), secondTarget && __copyProps(secondTarget, mod, "default"));
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var src_exports = {};
__export(src_exports, {
  Flow: () => import_flow.Flow,
  FlowServer: () => import_flow.FlowServer,
  GENKIT_CLIENT_HEADER: () => GENKIT_CLIENT_HEADER,
  GENKIT_VERSION: () => GENKIT_VERSION,
  GenkitError: () => import_error.GenkitError,
  defineFlow: () => import_flow.defineFlow,
  defineJsonSchema: () => import_schema.defineJsonSchema,
  defineSchema: () => import_schema.defineSchema,
  defineStreamingFlow: () => import_flow.defineStreamingFlow,
  getFlowAuth: () => import_auth.getFlowAuth,
  run: () => import_flow.run,
  z: () => import_zod.z
});
module.exports = __toCommonJS(src_exports);
var import_version = require("./__codegen/version.js");
var import_zod = require("zod");
__reExport(src_exports, require("./action.js"), module.exports);
var import_auth = require("./auth.js");
var import_error = require("./error.js");
var import_flow = require("./flow.js");
__reExport(src_exports, require("./flowTypes.js"), module.exports);
__reExport(src_exports, require("./plugin.js"), module.exports);
__reExport(src_exports, require("./reflection.js"), module.exports);
var import_schema = require("./schema.js");
__reExport(src_exports, require("./telemetryTypes.js"), module.exports);
__reExport(src_exports, require("./utils.js"), module.exports);
const GENKIT_VERSION = import_version.version;
const GENKIT_CLIENT_HEADER = `genkit-node/${GENKIT_VERSION} gl-node/${process.versions.node}`;
// Annotate the CommonJS export names for ESM import in node:
0 && (module.exports = {
  Flow,
  FlowServer,
  GENKIT_CLIENT_HEADER,
  GENKIT_VERSION,
  GenkitError,
  defineFlow,
  defineJsonSchema,
  defineSchema,
  defineStreamingFlow,
  getFlowAuth,
  run,
  z,
  ...require("./action.js"),
  ...require("./flowTypes.js"),
  ...require("./plugin.js"),
  ...require("./reflection.js"),
  ...require("./telemetryTypes.js"),
  ...require("./utils.js")
});
