// @genkit-ai/core/lib/reflection.js
"use strict";
var __create = Object.create;
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __getOwnPropSymbols = Object.getOwnPropertySymbols;
var __getProtoOf = Object.getPrototypeOf;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __propIsEnum = Object.prototype.propertyIsEnumerable;
var __defNormalProp = (obj, key, value) => key in obj ? __defProp(obj, key, { enumerable: true, configurable: true, writable: true, value }) : obj[key] = value;
var __spreadValues = (a, b) => {
  for (var prop in b || (b = {}))
    if (__hasOwnProp.call(b, prop))
      __defNormalProp(a, prop, b[prop]);
  if (__getOwnPropSymbols)
    for (var prop of __getOwnPropSymbols(b)) {
      if (__propIsEnum.call(b, prop))
        __defNormalProp(a, prop, b[prop]);
    }
  return a;
};
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toESM = (mod, isNodeMode, target) => (target = mod != null ? __create(__getProtoOf(mod)) : {}, __copyProps(
  // If the importer is in node compatibility mode or this is not an ESM
  // file that has been converted to a CommonJS file using a Babel-
  // compatible transform (i.e. "__esModule" has not been set), then set
  // "default" to the CommonJS "module.exports" for node compatibility.
  isNodeMode || !mod || !mod.__esModule ? __defProp(target, "default", { value: mod, enumerable: true }) : target,
  mod
));
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var __async = (__this, __arguments, generator) => {
  return new Promise((resolve, reject) => {
    var fulfilled = (value) => {
      try {
        step(generator.next(value));
      } catch (e) {
        reject(e);
      }
    };
    var rejected = (value) => {
      try {
        step(generator.throw(value));
      } catch (e) {
        reject(e);
      }
    };
    var step = (x) => x.done ? resolve(x.value) : Promise.resolve(x.value).then(fulfilled, rejected);
    step((generator = generator.apply(__this, __arguments)).next());
  });
};
var reflection_exports = {};
__export(reflection_exports, {
  ReflectionServer: () => ReflectionServer,
  RunActionResponseSchema: () => RunActionResponseSchema
});
module.exports = __toCommonJS(reflection_exports);
var import_express = __toESM(require("express"));
var import_promises = __toESM(require("fs/promises"));
var import_get_port = __toESM(require("get-port"));
var import_path = __toESM(require("path"));
var import_zod = __toESM(require("zod"));
var import_action = require("./action.js");
var import_index = require("./index.js");
var import_logging = require("./logging.js");
var import_schema = require("./schema.js");
var import_tracing = require("./tracing.js");
const RunActionResponseSchema = import_zod.default.object({
  result: import_zod.default.unknown().optional(),
  error: import_zod.default.unknown().optional(),
  telemetry: import_zod.default.object({
    traceId: import_zod.default.string().optional()
  }).optional()
});
const _ReflectionServer = class _ReflectionServer {
  constructor(registry, options) {
    /** Port the server is actually running on. This may differ from `options.port` if the original was occupied. Null if server is not running. */
    this.port = null;
    /** Express server instance. Null if server is not running. */
    this.server = null;
    /** Path to the runtime file. Null if server is not running. */
    this.runtimeFilePath = null;
    this.registry = registry;
    this.options = __spreadValues({
      port: 3100,
      bodyLimit: "30mb",
      configuredEnvs: ["dev"]
    }, options);
  }
  /**
   * Finds a free port to run the server on based on the original chosen port and environment.
   */
  findPort() {
    return __async(this, null, function* () {
      const chosenPort = this.options.port;
      const freePort = yield (0, import_get_port.default)({
        port: (0, import_get_port.makeRange)(chosenPort, chosenPort + 100)
      });
      if (freePort !== chosenPort) {
        import_logging.logger.warn(
          `Port ${chosenPort} is already in use, using next available port ${freePort} instead.`
        );
      }
      return freePort;
    });
  }
  /**
   * Starts the server.
   *
   * The server will be registered to be shut down on process exit.
   */
  start() {
    return __async(this, null, function* () {
      const server = (0, import_express.default)();
      server.use(import_express.default.json({ limit: this.options.bodyLimit }));
      server.use(function(req, res, next) {
        res.header("x-genkit-version", import_index.GENKIT_VERSION);
        next();
      });
      server.get("/api/__health", (_, response) => __async(this, null, function* () {
        yield this.registry.listActions();
        response.status(200).send("OK");
      }));
      server.get("/api/__quitquitquit", (_, response) => __async(this, null, function* () {
        import_logging.logger.debug("Received quitquitquit");
        response.status(200).send("OK");
        yield this.stop();
      }));
      server.get("/api/actions", (_, response, next) => __async(this, null, function* () {
        import_logging.logger.debug("Fetching actions.");
        try {
          const actions = yield this.registry.listActions();
          const convertedActions = {};
          Object.keys(actions).forEach((key) => {
            const action = actions[key].__action;
            convertedActions[key] = {
              key,
              name: action.name,
              description: action.description,
              metadata: action.metadata
            };
            if (action.inputSchema || action.inputJsonSchema) {
              convertedActions[key].inputSchema = (0, import_schema.toJsonSchema)({
                schema: action.inputSchema,
                jsonSchema: action.inputJsonSchema
              });
            }
            if (action.outputSchema || action.outputJsonSchema) {
              convertedActions[key].outputSchema = (0, import_schema.toJsonSchema)({
                schema: action.outputSchema,
                jsonSchema: action.outputJsonSchema
              });
            }
          });
          response.send(convertedActions);
        } catch (err) {
          const { message, stack } = err;
          next({ message, stack });
        }
      }));
      server.post("/api/runAction", (request, response, next) => __async(this, null, function* () {
        const { key, input } = request.body;
        const { stream } = request.query;
        import_logging.logger.debug(`Running action \`${key}\` with stream=${stream}...`);
        let traceId;
        try {
          const action = yield this.registry.lookupAction(key);
          if (!action) {
            response.status(404).send(`action ${key} not found`);
            return;
          }
          if (stream === "true") {
            const result = yield (0, import_action.runWithStreamingCallback)(
              (chunk) => {
                response.write(JSON.stringify(chunk) + "\n");
              },
              () => __async(this, null, function* () {
                return yield action.run(input);
              })
            );
            yield (0, import_tracing.flushTracing)();
            response.write(
              JSON.stringify({
                result: result.result,
                telemetry: {
                  traceId: result.telemetry.traceId
                }
              })
            );
            response.end();
          } else {
            const result = yield action.run(input);
            yield (0, import_tracing.flushTracing)();
            response.send({
              result: result.result,
              telemetry: {
                traceId: result.telemetry.traceId
              }
            });
          }
        } catch (err) {
          const { message, stack } = err;
          next({ message, stack, traceId });
        }
      }));
      server.get("/api/envs", (_, response) => __async(this, null, function* () {
        response.json(this.options.configuredEnvs);
      }));
      server.post("/api/notify", (request, response) => __async(this, null, function* () {
        const { telemetryServerUrl } = request.body;
        if (typeof telemetryServerUrl === "string") {
          (0, import_tracing.setTelemetryServerUrl)(telemetryServerUrl);
          import_logging.logger.debug(`Connected to telemetry server on ${telemetryServerUrl}`);
        }
        response.status(200).send("OK");
      }));
      server.use((err, req, res, next) => {
        import_logging.logger.error(err.stack);
        const error = err;
        const { message, stack } = error;
        const errorResponse = {
          code: import_action.StatusCodes.INTERNAL,
          message,
          details: {
            stack
          }
        };
        if (err.traceId) {
          errorResponse.details.traceId = err.traceId;
        }
        res.status(500).json(errorResponse);
      });
      this.port = yield this.findPort();
      this.server = server.listen(this.port, () => __async(this, null, function* () {
        import_logging.logger.debug(
          `Reflection server (${process.pid}) running on http://localhost:${this.port}`
        );
        _ReflectionServer.RUNNING_SERVERS.push(this);
        yield this.writeRuntimeFile();
      }));
    });
  }
  /**
   * Stops the server and removes it from the list of running servers to clean up on exit.
   */
  stop() {
    return __async(this, null, function* () {
      if (!this.server) {
        return;
      }
      return new Promise((resolve, reject) => __async(this, null, function* () {
        yield this.cleanupRuntimeFile();
        this.server.close((err) => __async(this, null, function* () {
          if (err) {
            import_logging.logger.error(
              `Error shutting down reflection server on port ${this.port}: ${err}`
            );
            reject(err);
          }
          const index = _ReflectionServer.RUNNING_SERVERS.indexOf(this);
          if (index > -1) {
            _ReflectionServer.RUNNING_SERVERS.splice(index, 1);
          }
          import_logging.logger.debug(
            `Reflection server on port ${this.port} has successfully shut down.`
          );
          this.port = null;
          this.server = null;
          resolve();
        }));
      }));
    });
  }
  /**
   * Writes the runtime file to the project root.
   */
  writeRuntimeFile() {
    return __async(this, null, function* () {
      try {
        const rootDir = yield findProjectRoot();
        const runtimesDir = import_path.default.join(rootDir, ".genkit", "runtimes");
        const date = /* @__PURE__ */ new Date();
        const time = date.getTime();
        const timestamp = date.toISOString();
        this.runtimeFilePath = import_path.default.join(
          runtimesDir,
          `${process.pid}-${time}.json`
        );
        const fileContent = JSON.stringify(
          {
            id: process.env.GENKIT_RUNTIME_ID || process.pid.toString(),
            pid: process.pid,
            reflectionServerUrl: `http://localhost:${this.port}`,
            timestamp
          },
          null,
          2
        );
        yield import_promises.default.mkdir(runtimesDir, { recursive: true });
        yield import_promises.default.writeFile(this.runtimeFilePath, fileContent, "utf8");
        import_logging.logger.debug(`Runtime file written: ${this.runtimeFilePath}`);
      } catch (error) {
        import_logging.logger.error(`Error writing runtime file: ${error}`);
      }
    });
  }
  /**
   * Cleans up the port file.
   */
  cleanupRuntimeFile() {
    return __async(this, null, function* () {
      if (!this.runtimeFilePath) {
        return;
      }
      try {
        const fileContent = yield import_promises.default.readFile(this.runtimeFilePath, "utf8");
        const data = JSON.parse(fileContent);
        if (data.pid === process.pid) {
          yield import_promises.default.unlink(this.runtimeFilePath);
          import_logging.logger.debug(`Runtime file cleaned up: ${this.runtimeFilePath}`);
        }
      } catch (error) {
        import_logging.logger.error(`Error cleaning up runtime file: ${error}`);
      }
    });
  }
  /**
   * Stops all running reflection servers.
   */
  static stopAll() {
    return __async(this, null, function* () {
      return Promise.all(
        _ReflectionServer.RUNNING_SERVERS.map((server) => server.stop())
      );
    });
  }
};
/** List of all running servers needed to be cleaned up on process exit. */
_ReflectionServer.RUNNING_SERVERS = [];
let ReflectionServer = _ReflectionServer;
function findProjectRoot() {
  return __async(this, null, function* () {
    let currentDir = process.cwd();
    while (currentDir !== import_path.default.parse(currentDir).root) {
      const packageJsonPath = import_path.default.join(currentDir, "package.json");
      try {
        yield import_promises.default.access(packageJsonPath);
        return currentDir;
      } catch (e) {
        currentDir = import_path.default.dirname(currentDir);
      }
    }
    throw new Error("Could not find project root (package.json not found)");
  });
}
if (typeof module !== "undefined" && "hot" in module) {
  module.hot.accept();
  module.hot.dispose(() => __async(void 0, null, function* () {
    import_logging.logger.debug("Cleaning up reflection server(s) before module reload...");
    yield ReflectionServer.stopAll();
  }));
}
// Annotate the CommonJS export names for ESM import in node:
0 && (module.exports = {
  ReflectionServer,
  RunActionResponseSchema
});
