// @genkit-ai/tools-common/lib/esm/utils/utils.js
import * as fs from 'fs/promises';
import * as path from 'path';
export async function findProjectRoot() {
    let currentDir = process.cwd();
    while (currentDir !== path.parse(currentDir).root) {
        const packageJsonPath = path.join(currentDir, 'package.json');
        const goModPath = path.join(currentDir, 'go.mod');
        try {
            const [packageJsonExists, goModExists] = await Promise.all([
                fs
                    .access(packageJsonPath)
                    .then(() => true)
                    .catch(() => false),
                fs
                    .access(goModPath)
                    .then(() => true)
                    .catch(() => false),
            ]);
            if (packageJsonExists || goModExists) {
                return currentDir;
            }
        }
        catch {
        }
        currentDir = path.dirname(currentDir);
    }
    throw new Error('Could not find project root');
}
export async function findRuntimesDir(projectRoot) {
    const root = projectRoot ?? (await findProjectRoot());
    return path.join(root, '.genkit', 'runtimes');
}
export async function findServersDir(projectRoot) {
    const root = projectRoot ?? (await findProjectRoot());
    return path.join(root, '.genkit', 'servers');
}
export function projectNameFromGenkitFilePath(filePath) {
    const parts = filePath.split('/');
    const basePath = parts
        .slice(0, Math.max(parts.findIndex((value) => value === '.genkit'), 0))
        .join('/');
    return basePath === '' ? 'unknown' : path.basename(basePath);
}
export async function detectRuntime(directory) {
    const files = await fs.readdir(directory);
    for (const file of files) {
        const filePath = path.join(directory, file);
        const stat = await fs.stat(filePath);
        if (stat.isFile() && (path.extname(file) === '.go' || file === 'go.mod')) {
            return 'go';
        }
    }
    try {
        await fs.access(path.join(directory, 'package.json'));
        return 'nodejs';
    }
    catch {
        return undefined;
    }
}
export async function checkServerHealth(url) {
    try {
        const response = await fetch(`${url}/api/__health`);
        return response.status === 200;
    }
    catch (error) {
        if (error instanceof Error &&
            error.cause.code === 'ECONNREFUSED') {
            return false;
        }
    }
    return true;
}
export async function waitUntilHealthy(url, maxTimeout = 10000) {
    const startTime = Date.now();
    while (Date.now() - startTime < maxTimeout) {
        try {
            const response = await fetch(`${url}/api/__health`);
            if (response.status === 200) {
                return true;
            }
        }
        catch (error) {
        }
        await new Promise((resolve) => setTimeout(resolve, 500));
    }
    return false;
}
export async function waitUntilUnresponsive(url, maxTimeout = 10000) {
    const startTime = Date.now();
    while (Date.now() - startTime < maxTimeout) {
        try {
            const health = await fetch(`${url}/api/__health`);
        }
        catch (error) {
            if (error instanceof Error &&
                error.cause.code === 'ECONNREFUSED') {
                return true;
            }
        }
        await new Promise((resolve) => setTimeout(resolve, 500));
    }
    return false;
}
export async function retriable(fn, opts) {
    const maxRetries = opts.maxRetries ?? 3;
    const delayMs = opts.delayMs ?? 0;
    let attempt = 0;
    while (true) {
        try {
            return await fn();
        }
        catch (e) {
            if (attempt >= maxRetries - 1) {
                throw e;
            }
            if (delayMs > 0) {
                await new Promise((r) => setTimeout(r, delayMs));
            }
        }
        attempt++;
    }
}
