// @genkit-ai/tools-common/lib/cjs/types/document.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.DocumentDataSchema = exports.PartSchema = exports.MediaPartSchema = exports.TextPartSchema = void 0;
const zod_1 = __importDefault(require("zod"));
const EmptyPartSchema = zod_1.default.object({
    text: zod_1.default.never().optional(),
    media: zod_1.default.never().optional(),
});
exports.TextPartSchema = EmptyPartSchema.extend({
    text: zod_1.default.string(),
});
exports.MediaPartSchema = EmptyPartSchema.extend({
    media: zod_1.default.object({
        contentType: zod_1.default.string().optional(),
        url: zod_1.default.string(),
    }),
});
exports.PartSchema = zod_1.default.union([exports.TextPartSchema, exports.MediaPartSchema]);
exports.DocumentDataSchema = zod_1.default.object({
    content: zod_1.default.array(exports.PartSchema),
    metadata: zod_1.default.record(zod_1.default.string(), zod_1.default.any()).optional(),
});
